services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java

```diff
@@ -945,6 +976,56 @@ protected Response exchangeClientToClient(UserModel targetUser, UserSessionModel
         return cors.builder(Response.ok(res, MediaType.APPLICATION_JSON_TYPE)).build();
     }
 
+    protected Response exchangeClientToSAML2Client(UserModel targetUser, UserSessionModel targetUserSession, String requestedTokenType,
+                                                  ClientModel targetClient, String audience, String scope) {
+        // Create authSession with target SAML 2.0 client and authenticated user
+        LoginProtocolFactory factory = (LoginProtocolFactory) session.getKeycloakSessionFactory()
+                .getProviderFactory(LoginProtocol.class, SamlProtocol.LOGIN_PROTOCOL);
+        SamlService samlService = (SamlService) factory.createProtocolEndpoint(realm, event);
+        ResteasyProviderFactory.getInstance().injectProperties(samlService);
+        AuthenticationSessionModel authSession = samlService.getOrCreateLoginSessionForIdpInitiatedSso(session, realm,
+                targetClient, null);
+        if (authSession == null) {
+            logger.error("SAML assertion consumer url not set up");
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_CLIENT, "Client requires assertion consumer url set up", Response.Status.BAD_REQUEST);
+        }
+
+        authSession.setAuthenticatedUser(targetUser);
+
+        event.session(targetUserSession);
+
+        AuthenticationManager.setClientScopesInSession(authSession);
+        ClientSessionContext clientSessionCtx = TokenManager.attachAuthenticationSession(this.session, targetUserSession,
+                authSession);
+
+        updateUserSessionFromClientAuth(targetUserSession);
+
+        // Create SAML 2.0 Assertion Response
+        SamlClient samlClient = new SamlClient(targetClient);
+        SamlProtocol samlProtocol = new TokenExchangeSamlProtocol(samlClient).setEventBuilder(event).setHttpHeaders(headers).setRealm(realm)
+                .setSession(session).setUriInfo(session.getContext().getUri());
+
+        Response samlAssertion = samlProtocol.authenticated(authSession, targetUserSession, clientSessionCtx);
+        if (samlAssertion.getStatus() != 200) {
+            throw new CorsErrorResponseException(cors, OAuthErrorException.INVALID_REQUEST, "Can not get SAML 2.0 token", Response.Status.BAD_REQUEST);
+        }
+        String xmlString = (String) samlAssertion.getEntity();
+        String encodedXML = Base64Url.encode(xmlString.getBytes(GeneralConstants.SAML_CHARSET));
+
+        int assertionLifespan = samlClient.getAssertionLifespan();
+
+        AccessTokenResponse res = new AccessTokenResponse();
```

Comment:
Could you include `issued_token_type` and set it to `urn:ietf:params:oauth:token-type:saml2`. See https://tools.ietf.org/id/draft-ietf-oauth-token-exchange-19.html#rfc.section.3.

I guess that would mean creating a `TokenExchangeResponse` extending `AccessTokenResponse` ? Reason being that we use that new class only for token exchange and don't touch the oauth2 part. Also easier to support additional parameters in the future that could be specific of exchange.

Reply:
@pedroigor 

> Could you include issued_token_type and set it to `urn:ietf:params:oauth:token-type:saml2`

It's already included in the following code through `AccessTokenResponse#setOtherClaims` method:
https://github.com/keycloak/keycloak/pull/7064/files#diff-ec5f9626f53934cdc3ad494e198f1902R1021

But there is no assertion in the test code. I'll add the assertion.

> I guess that would mean creating a `TokenExchangeResponse` extending `AccessTokenResponse`

- Create `TokenExchangeResponse` extending `AccessTokenResponse` as new class
- Add `issuedTokenType` field and set/get method into the new class
- Fix the existing token exchange code for OIDC client and external clients (implementation of `ExchangeTokenToIdentityProviderToken`)to use the new response class

Is the above fix OK?

Also, I noticed the current token exchange code doesn't follow the spec.

- Exchange for OIDC client doesn't include `issued_token_type` claim.
- Refresh token for OIDC client isn't stored in `access_token` claim when `requested_token_type` is refresh token type. 
https://github.com/keycloak/keycloak/blob/6dde131609551621eba102d71c5897a350d7ce64/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L930-L933
- ID token is issued by the `scope` parameter. Should we issue the id token when `requested_token_type` is id token type?
https://github.com/keycloak/keycloak/blob/6dde131609551621eba102d71c5897a350d7ce64/services/src/main/java/org/keycloak/protocol/oidc/endpoints/TokenEndpoint.java#L935-L938
- ID token for external client isn't stored in `access_token` claim when `requested_token_type` is id token type. 
https://github.com/keycloak/keycloak/blob/7efaf9869acbb328f9dd8b159460cec7de6bc010/services/src/main/java/org/keycloak/broker/oidc/AbstractOAuth2IdentityProvider.java#L225-L230

Should I fix those issues in this pull request too?

Reply:
Yeah, we are missing some bits of the specs. I really missed the part you are setting the `issued_token_type`.

Don't worry about fixing them now. Let's do it in a separate and specific PR.

Reply:
All right, I'll fix `DocumentUtil` and test case.